src/db-service/shopping_cart.js

```diff
@@ -104,3 +104,10 @@ exports.checkoutCart = async products => {
   });
   return Promise.all(queries);
 };
+
+exports.productInventory = async (id) => {
+  const rows = await knex("products")
+    .select("inventory_count")
+    .where({ id });
+  return rows[0] || false;
+};
```

Comment:
One thing I want to add is that the connections are disposed on the knex side but the database is never notified about this due to communication error. The inactive sessions/connections stay on the database side. Not sure if the database eventually kills them though depending on the settings. 
This flow can be represented as: "client grabbed a connection, hit an exception, propagated over and out of the scope of the connection handle from (a), so it can never give it back - it lost the track of connection.
 
Is there any way you know that can help this issue or any generally related experience?

Reply:
Note that this should also be an extremely rare case, supplying such a long passphrase does not make sense since the entropy of the passphrase would far exceed the entropy of the derived decryption key.

Reply:
In that case, I'd probably ask the user for the IV size; if they know enough about crypto to pick a cipher for a specific application, then they should also be able to supply its parameters. What would you do? For example, what should the IV size for AES-CCM be? pycryptodome, a popular crypto library, says 11 bytes, a reasonably strong choice. OpenSSL only uses 7 bytes, a much weaker choice, which has the advantage of permitting larger messages to be encrypted. Similar problems arise with authentication tag lengths etc. We would have to decide without knowing what the user is going to do with the cipher.

Also, there are API differences for different ciphers. All of the following statements are true for some supported ciphers but not for all:

supports streaming behavior
has a unique valid IV length
supports AEAD
has a maximum message size
is resistant against padding oracle attacks
its output length is a multiple of the block size
only works if message integrity is guaranteed by using an authentication tag
needs to know the length of the message in advance
designed for message encryption
designed for symmetric key wrapping
If you let users pick any valid cipher, you end up with an arbitrary subset of these attributes, and your application needs to somehow consider all of these points and their combinations.